### ISSUE TITLE
DNF local Plugin

### ISSUE THREAD

**Issue by j-mracek** [rpm-software-management/dnf5#926]
Plugin mostly important only for rawhide users (unstable branch without possibility to downgrade).

Do we want to implement it?

**Comment by gjohnz** [comment]
I'd like to see python3-dnf-plugin-local included. It is a very useful plugin IMO. I've been using it since Fedora 29.

**Comment by buckaroogeek** [comment]
I use the dnf local plugin extensively. My internet connection is wireless (rural area) and bandwidth is expensive. The local plugin allows me to easily re-use downloaded packages in multiple virtual and physical machines that run Fedora. Like @gjohnz, I have been using this since Fedora 33 (more or less).

**Comment by 10leej** [comment]
The lack of support for dnf5 is what blocks upgrading to fedora 41 for me. for many of the same reasons @buckaroogeek mentioned.

**Comment by jsullivan3** [comment]
I put together [this package](https://github.com/jsullivan3/dnf-local-cache-actions/releases/tag/v0.1-1) based on `libdnf5-plugin-actions` functionality.  It seems to be solving the issue for me, and can read the configuration file from the `python3-dnf-plugin-local` package.  Hopefully you find it helpful too!

**Comment by gjohnz** [comment]
I do get a permission error when using **dnf repolist** unless I use _sudo_
```

gjohnson@fedora=> dnf repolist
/usr/sbin/dnf-local-cache-actions: line 16: /var/log/dnf-local-cache-actions.log: Permission denied
/usr/sbin/dnf-local-cache-actions: line 21: /var/log/dnf-local-cache-actions.log: Permission denied
repo id                                                 repo name                                                            
fedora                                                  Fedora 41 - x86_64                                                   
fedora-cisco-openh264                  Fedora 41 openh264 (From Cisco) - x86_64                             
updates                                                 Fedora 41 - x86_64 - Updates
```

**Comment by jsullivan3** [comment]
Oh cool!  Thank you very much @gjohnz!  I see what's going on there - the log file is owned by `root`, so a non-root user will be unable to write to it.  Probably tied to the logging issue I created earlier.  There seems to be another issue with the `/etc/dnf/plugins/local.conf` file being readable by only root, so even if the plugin can write to the log file, it won't be able to read the configuration.

The following workaround will make it usable for non-root users, though it unfortunately involves the rather yucky hack of making the log file world-writable:

```
sudo chmod 0644 /etc/dnf/plugins/local.conf
sudo chmod 0666 /var/log/dnf-local-cache-actions.log
```

With these two changes, I'm able to successfully list repositories as a non-root user.

Again, thank you for giving it a try and telling me about the issue!

**Comment by gjohnz** [comment]
I've tested this several times now and I've concluded that previously downloaded rpms are not being pulled from the local repository, they are simply being downloaded again. Might I be missing something?

@jsullivan3 again thank you for your time and effort with this "plugin".

**Comment by jsullivan3** [comment]
My pleasure, @gjohnz!  I just created [an issue](https://github.com/jsullivan3/dnf-local-cache-actions/issues/3) in the dnf-local-cache-actions repository to track this behavior.  Would you be willing to continue the discussion there?  Thank you again for your patience and willingness to test this plugin.

**Comment by BenjiWiebe** [comment]
I've been wanting to set up dnf-local on NFS, as I have multiple Fedora machines on the LAN, but not a very fast WAN connection. I set it up now after upgrading to F41, and am very disappointed to see that dnf5 does not support it.

**Comment by vorpalfish** [comment]
I also hope this gets implemented eventually; I have 5 fedora machines to update,  so local plugin has been dropping internet usage to 20% of what it otherwise would be. I've been forcing the use of dnf4 since the upgrade to F41 so I can keep using it, but I imagine that won't be around forever.

**Comment by BenjiWiebe** [comment]
> I also hope this gets implemented eventually; I have 5 fedora machines to update, so local plugin has been dropping internet usage to 20% of what it otherwise would be. I've been forcing the use of dnf4 since the upgrade to F41 so I can keep using it, but I imagine that won't be around forever.

I did set up the dnf-local-cache-actions from @jsullivan3 mentioned earlier in this thread, and it is working fine so far with dnf5. It would be nice to have it builtin, but I'm happy with what I've got now too.

**Comment by indyprime** [comment]
I have also used this for a few years, and keep using dnf3 since dnf5 doesn't support the plugin. Saves bandwidth for multiple physical and virtual machines, and perhaps preserves a bit of privacy too.

I do see dnf5 as being more efficient, but the local cache wins out. Any idea whether this will be officially added to dnf5?

**Comment by buckaroogeek** [comment]
The dnf documentation does not list the local plugin (https://dnf5.readthedocs.io/en/latest/dnf5.8.html#plugin-commands) as available. I suspect that is still works only for dnf v4. There is another similar tool available for dnf 5 but not directly supported by the dnf team as far as I can tell.

**Comment by indyprime** [comment]
Yes, the plugin doesn't work in dnf5, and ignores the local package cache. All I have to do is replace the /usr/bin/dnf symlink to point to /usr/bin/dnf4 (which itself is just a symlink to /usr/bin/dnf-3), and the plugin works again.

This fix is only temporary, since every dnf update points the symlink back to /usr/bin/dnf5, but that's easily replaced again. Of course, this would be better if dnf5 supported the plugin.
